spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -0,0 +1,164 @@
+          matchedConditions = actions.map(resolveClauseCondition(_, joinPlan)),
+          matchedOutputs = actions.map(actionOutput(_, targetOutputCols, joinPlan)),
+          notMatchedConditions = notActions.map(resolveClauseCondition(_, joinPlan)),
+          notMatchedOutputs = notActions.map(actionOutput(_, targetOutputCols, joinPlan)),
+          targetOutput = resolveExprs(targetOutputCols :+ Literal(false), joinPlan),
+          joinedAttributes = joinPlan.output
+        )
+
+        val mergePlan = MergeInto(mergeIntoProcessor, target, joinPlan)
+        val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+        ReplaceData(target, batchWrite, mergePlan)
+    }
+  }
+
+  private def buildScanPlan(
+      table: Table,
+      output: Seq[AttributeReference],
+      mergeBuilder: MergeBuilder,
+      prunedTargetPlan: LogicalPlan): LogicalPlan = {
+
+    val scanBuilder = mergeBuilder.asScanBuilder
+    val scan = scanBuilder.build()
+    val scanRelation = DataSourceV2ScanRelation(table, scan, output)
+
+    scan match {
+      case _: SupportsFileFilter =>
+        val matchingFilePlan = buildFileFilterPlan(prunedTargetPlan)
+        val dynamicFileFilter = DynamicFileFilter(scanRelation, matchingFilePlan)
+        dynamicFileFilter
+      case _ =>
+        scanRelation
+    }
+  }
+
+  private def newWriteInfo(schema: StructType): LogicalWriteInfo = {
+    val uuid = UUID.randomUUID()
+    LogicalWriteInfoImpl(queryId = uuid.toString, schema, CaseInsensitiveStringMap.empty)
+  }
+
+  private def buildFileFilterPlan(prunedTargetPlan: LogicalPlan): LogicalPlan = {
+    val fileAttr = findOutputAttr(prunedTargetPlan, FILE_NAME_COL)
+    Aggregate(Seq(fileAttr), Seq(fileAttr), prunedTargetPlan)
+  }
+
+  private def findOutputAttr(plan: LogicalPlan, attrName: String): Attribute = {
+    val resolver = SQLConf.get.resolver
+    plan.output.find(attr => resolver(attr.name, attrName)).getOrElse {
+      throw new AnalysisException(s"Cannot find $attrName in ${plan.output}")
+    }
+  }
+
+  private def resolveExprs(exprs: Seq[Expression], plan: LogicalPlan): Seq[Expression] = {
+    val spark = SparkSession.active
+    exprs.map { expr => resolveExpressionInternal(spark, expr, plan) }
+  }
+
+  def getTargetOutputCols(target: DataSourceV2Relation): Seq[NamedExpression] = {
+    target.schema.map { col =>
+      target.output.find(attr => SQLConf.get.resolver(attr.name, col.name)).getOrElse {
+        Alias(Literal(null, col.dataType), col.name)()
+      }
+    }
+  }
+
+  def actionOutput(clause: MergeAction,
+                   targetOutputCols: Seq[Expression],
+                   plan: LogicalPlan): Seq[Expression] = {
+    val exprs = clause match {
+      case u: UpdateAction =>
+        u.assignments.map(_.value) :+ Literal(false)
+      case _: DeleteAction =>
+        targetOutputCols :+ Literal(true)
+      case i: InsertAction =>
+        i.assignments.map(_.value) :+ Literal(false)
```

Comment:
As Anton noted, there needs to be validation that the assignments here (for both inserts and updates) match up with the `targetOutputCols`.